dask_expr/io/io.py

```diff
@@ -36,7 +36,7 @@ class FromGraph(IO):
     conversion from legacy dataframes.
     """
 
-    _parameters = ["layer", "_meta", "divisions", "_name"]
+    _parameters = ["layer", "_meta", "divisions", "keys", "name_prefix"]
```

Comment:
The `FromGraph` thing got a little odd. For context, this thing is used in two ways

1. As a wrapper for futures when persist is called
2. As a wrapper for legacy dataframes

The previous implementation accepted the `_name` as an input argument. For the persisted dataframe, this was the name of the original expression. For the wrapped one, it is the name of the legacy dataframe.

The problem is that the legacy dataframe does not have the same strict uniqueness guarantees as the expressions do so setting duplicate names is very easy. In fact, our tests where doing just that! (and still are). This caused the `Expr.__new__` to deduplicate and effectively ignore the second dataframe... oops.

For persist it is also a little odd since if the `FromGraph` expression is inheriting the exact name of it's ancestor, there exist now two expressions of a different type with the same name. This is odd.


Effectively, with this chosen singleton implementation, setting the `_name` explicitly instead of calculating it using a hash is a cardinal sin and will cause all sorts of weird things to happen.

Now, this can be fixed but that has a rather big caveat. If I have to redefine the name of the expression, I actually also have to rewrite the graph! Many places in dask-expr are (imo incorrectly) assuming that output key names of a dataframe layer/expression are universally built as `(df._name, i)` and are hard coding this when implementing their own layer (instead of iterating over `i`, iterating over `df.__dask_keys__()` would maintain the abstraction). This rewrite adds effectively another layer of keys. In reality this is really ugly since when computing something on top of a persisted dataframe, there will always be this dummy key in between.

Alternatively, I could make the singleton deduplication type aware to give the `FromGraph` thing an excuse to overwrite the name. However, if we truly stick with singletons that are based on the name, I would prefer the name to actually be unique which required all implementations to stop hard coding keys of another expression/dataframe and iterate properly over the `__dask_keys__`

Reply:
> Now, this can be fixed but that has a rather big caveat. If I have to redefine the name of the expression, I actually also have to rewrite the graph! Many places in dask-expr are (imo incorrectly) assuming that output key names of a dataframe layer/expression are universally built as (df._name, i) and are hard coding this when implementing their own layer (instead of iterating over i, iterating over df.__dask_keys__() would maintain the abstraction). This rewrite adds effectively another layer of keys. In reality this is really ugly since when computing something on top of a persisted dataframe, there will always be this dummy key in between.

This is a good point, we should fix this instead of relying on names

Reply:
I will investigate if it is possible to just use `__dask_keys__` everywhere but I'd prefer doing this in a follow up

Reply:
Yes totally agree, this should definitely be a follow up (also doesn't have to be you, I could pick this up as well)

Reply:
I got started on this, see

https://github.com/dask-contrib/dask-expr/commit/c3c01ed57d7986d1ad8c89f311a2bce2e453d08c

That commit replaces all occurrences that match `\((\w+(_\w+)?)(?<!self)\._name,` but something appears to be still missing. It is possible and not as ugly as I thought it would be